lib/puppet/provider/cs_stonith/pcs.rb

```diff
@@ -0,0 +1,209 @@
+require 'pathname'
+require Pathname.new(__FILE__).dirname.dirname.expand_path + 'pacemaker'
+
+Puppet::Type.type(:cs_stonith).provide(:pcs, :parent => Puppet::Provider::Pacemaker) do
+  desc 'A provider for creating fence devices to perform STONITH operations
+	should the cluster find itself in trouble.
+        Here we manage the creation and deletion of STONITH fence devices
+        We will accept a hash for what Corosync calls
+        operations and device_options.  A hash is used instead of constucting a
+        better model since these values can be almost anything.'
+
+  commands :pcs => 'pcs'
+
+  defaultfor :operatingsystem => [:fedora, :centos, :redhat]
+  # given an XML element containing some <nvpair>s, return a hash. Return an
+  # empty hash if `e` is nil.
+  def self.nvpairs_to_hash(e)
+    return {} if e.nil?
+
+    hash = {}
+    e.each_element do |i|
+      hash[(i.attributes['name'])] = i.attributes['value']
+    end
+
+    hash
+  end
+
+  # given an XML element (a <primitive> from cibadmin), produce a hash suitible
+  # for creating a new provider instance.
+  def self.element_to_hash(e)
+    hash = {
+      :primitive_type           => e.attributes['type'],
+      :name                     => e.attributes['id'].to_sym,
+      :ensure                   => :present,
+      :provider                 => self.name,
+      :device_options           => nvpairs_to_hash(e.elements['instance_attributes']),
+      :operations               => {},
+      :existing_resource        => :true,
+      :existing_primitive_type  => e.attributes['type'],
+      :existing_operations      => {},
+    }
+
+    if ! e.elements['operations'].nil?
+      e.elements['operations'].each_element do |o|
+        valids = o.attributes.reject do |k,v| k == 'id' end
+        if ! valids['role'].nil?
+          name = valids['name']
+          name << ":"
+          name << valids['role']
+        else
+          name = valids['name']
+        end
+        hash[:operations][name] = {}
+        valids.each do |k,v|
+          hash[:operations][name][k] = v if k != 'name' and k != 'role'
+        end
+        hash[:existing_operations] = hash[:operations].dup
```

Comment:
This was a style copied from https://github.com/puppet-community/puppet-corosync/blob/master/lib/puppet/provider/cs_primitive/pcs.rb#L52